src/wasm/wasm-binary.cpp

```diff
@@ -1242,6 +1239,44 @@ void WasmBinaryWriter::writeSourceMapProlog() {
     }
   }
 
+  // Remove unused function names from 'names' field.
+  if (!wasm->debugInfoSymbolNames.empty()) {
+    std::vector<std::string> newSymbolNames;
+    std::unordered_map<Index, Index> oldToNewIndex;
+
+    // Collect all used symbol name indexes.
+    for (auto& func : wasm->functions) {
+      for (auto& pair : func->debugLocations) {
+        if (pair.second && pair.second->symbolNameIndex) {
+          uint32_t oldIndex = *pair.second->symbolNameIndex;
+          assert(oldIndex < wasm->debugInfoSymbolNames.size());
+          oldToNewIndex[oldIndex] = 0; // placeholder
+        }
+      }
+    }
+
+    // Create the new list of names and the mapping from old to new indices.
+    uint32_t newIndex = 0;
+    for (auto& pair : oldToNewIndex) {
```

Comment:
This is on an unordered map, so the indexing may end up nondeterministic, I worry?